packages/adapter-pg/src/pg-types.ts

```diff
@@ -0,0 +1,226 @@
+import { ok, type Query, type Result } from '@prisma/driver-adapter-utils'
+import type pg from 'pg'
+
+/**
+ * PostgreSQL array column types (not defined in `import('pg').types.builtins`).
+ *
+ * See the semantics of each of this code in:
+ *   https://github.com/postgres/postgres/blob/master/src/include/catalog/pg_type.dat
+ */
+export const ArrayColumnType = {
+  BIT_ARRAY: 1561,
+  BOOL_ARRAY: 1000,
+  BYTEA_ARRAY: 1001,
+  BPCHAR_ARRAY: 1014,
+  CHAR_ARRAY: 1002,
+  CIDR_ARRAY: 651,
+  DATE_ARRAY: 1182,
+  FLOAT4_ARRAY: 1021,
+  FLOAT8_ARRAY: 1022,
+  INET_ARRAY: 1041,
+  INT2_ARRAY: 1005,
+  INT4_ARRAY: 1007,
+  INT8_ARRAY: 1016,
+  JSONB_ARRAY: 3807,
+  JSON_ARRAY: 199,
+  MONEY_ARRAY: 791,
+  NUMERIC_ARRAY: 1231,
+  OID_ARRAY: 1028,
+  TEXT_ARRAY: 1009,
+  TIMESTAMP_ARRAY: 1115,
+  TIME_ARRAY: 1183,
+  UUID_ARRAY: 2951,
+  VARBIT_ARRAY: 1563,
+  VARCHAR_ARRAY: 1015,
+  XML_ARRAY: 143,
+}
+
+/**
+ * Maximal OID of system catalog entries
+ */
+export const maxSystemCatalogOID = 9999
+
+/**
+ * A PostgreSQL type representation with both OID and optional introspected type name.
+ */
+export type PgType = {
+  readonly id: number
+  readonly name?: string
+}
+
+/**
+ * Postgres types cache pluggable into the client. All connections to the same database should
+ * preferably use the same instance to avoid duplicating requests.
+ */
+export class PgTypesCache {
+  // cache the promises and not the results to avoid concurrent requests for the same type
+  #cachedRequests = new Map<number, Promise<Result<PgType>>>()
+
+  constructor(
+    private client: {
+      performIO(query: Query): Promise<Result<pg.QueryArrayResult<unknown[]>>>
+    },
+  ) {}
+
+  typeById(id: number): Promise<Result<PgType>> {
+    let req = this.#cachedRequests.get(id)
+
+    if (req === undefined) {
+      req = this.#loadType(id)
+      this.#cachedRequests.set(id, req)
+    }
+
+    return req
+  }
+
+  async #loadType(id: number): Promise<Result<PgType>> {
+    if (id <= maxSystemCatalogOID && knownSystemTypeNames[id]) {
+      return ok({
+        id,
+        name: knownSystemTypeNames[id],
+      })
+    }
+
+    const result = (
+      await this.client.performIO({
+        sql: 'SELECT typname FROM pg_catalog.pg_type WHERE oid = $1',
```

Comment:
Hum, this impacts coldstart performance. In order to support citext (or any other OID added by an extension) we do need to query the catalog indeed, however, I think we can do less queries provided that each user query selects various columns, and we have information about the oids for such columns as the user query is returned.  The algorithm will follow:

* user query returns oid set `O = { o1, o2, .., on. }`
* we look in the cache for individual oid type names, `oi | i belong {1..n}` was already in the cache from a previous query. 
* we query O - oi, and cache the results. individually. (Not their promises, tho)

What do you think? Will that improve cold start?

Reply:
If the schema contains 4 tables each with 5 columns. And user queries are "SELECT * FROM...", instead of 20 roundtrips for the enums this would lead to 4 roundtrips.

Reply:
That makes sense, we could indeed make less queries by looking up multiple OIDs at once.